docs/src/snap/howto/two-node-ha.md

```diff
@@ -0,0 +1,426 @@
+```
+
+```
+[Unit]
+Description=K8s service wrapper handling Dqlite recovery for two-node HA setups.
+After=network.target pacemaker.service
+
+[Service]
+User=ubuntu
+Group=ubuntu
+Type=oneshot
+ExecStart=/bin/bash /snap/k8s/current/k8s/hack/two-node-ha.sh start_service
+ExecStop=/bin/bash sudo snap stop k8s
+RemainAfterExit=true
+
+[Install]
+WantedBy=multi-user.target
+```
+
+```{note}
+The ``two-node-ha.sh start_service`` command used by the service wrapper
+automatically detects the expected Dqlite role based on the DRBD state.
+It then takes the necessary steps to bootstrap the Dqlite state directories,
+synchronize with the peer node (if available) and recover the database.
+```
+
+When a DRBD failover occurs, the ``two-node-ha-k8s`` service needs to be
+restarted. To accomplish this,, we are going to define a separate service that
+will be invoked by Pacemaker. Create a file called
+``/etc/systemd/system/two-node-ha-k8s-failover.service`` containing the
+following:
+
+```
+[Unit]
+Description=Managed by Pacemaker, restarts two-node-ha-k8s on failover.
+After=network.target home-ubuntu-workspace.mount
+
+[Service]
+Type=oneshot
+ExecStart=systemctl restart two-node-ha-k8s
+RemainAfterExit=true
+```
+
+Reload the systemd configuration and set ``two-node-ha-k8s`` to start
+automatically. Notice that ``two-node-ha-k8s-failover`` must not be configured
+to start automatically, but instead is going to be managed through Pacemaker.
+
+```
+sudo systemctl enable two-node-ha-k8s
+sudo systemctl daemon-reload
+```
+
+Make sure that both nodes have been configured using the above steps before
+moving forward.
+
+### Automating the failover procedure
+
+Define a new Pacemaker resource that will invoke the
+``two-node-ha-k8s-failover`` service when a DRBD failover occurs.
+
+```
+sudo crm configure <<EOF
+primitive ha_k8s_failover_service systemd:two-node-ha-k8s-failover op start interval=0 timeout=120 op stop interval=0 timeout=30
+order failover_after_fs mandatory: fs_res:start ha_k8s_failover_service:start
+colocation fs_failover_colo INFINITY: fs_res ha_k8s_failover_service
+commit
+show
+quit
+EOF
+```
+
+Once the setup is complete on both nodes, start the two-node HA k8s service on
+each node:
+
+```
+sudo systemctl start two-node-ha-k8s
+```
+
+## Troubleshooting
+
+Here are some potential problems that may affect two-node HA clusters and how
+to address them.
+
+### Dqlite recovery failing because of unexpected data segments
+
+Dqlite recovery may fail if there are data segments past the latest snapshot.
+
+```
+Error: failed to recover k8s-dqlite, error: k8s-dqlite recovery failed, error:
+recover failed with error code 1, error details: raft_recover(): io:
+closed segment 0000000000002369-0000000000002655 is past last snapshot
+snapshot-2-2048-642428, pre-recovery backup:
+/var/snap/k8s/common/recovery-k8s-dqlite-2024-09-05T082644Z-pre-recovery.tar.gz
+```
+
+Remove the offending segments and restart the ``two-node-ha-k8s`` service.
```

Comment:
As a user I would wonder if it is save to remove this segment. Should we add an explanation? This is the data that has not been replicated to more than half of the voters and can therefore be safely removed?

Reply:
In our scenario, Dqlite doesn't replicate data at all, which is why we need DRBD. Removing data segments could lead to data loss, but there's no other documented way of recovering the cluster.

To be honest, the dqlite documentation is rather scarce and doesn't even mention this. I had to rely on the LXD documentation as well as the comments on GitHub issues:

* https://documentation.ubuntu.com/lxd/en/latest/howto/dqlite_troubleshoot/#interventions
* https://github.com/canonical/dqlite/issues/309#issuecomment-854736238

Reply:
I see, can we add a warning similar to the one in LXD's docs please?